internal/mirror/database/mirror.go

```diff
@@ -0,0 +1,223 @@
+// Copyright 2020 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package database is a database interface for mirror settings.
+package database
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/google/exposure-notifications-server/internal/database"
+	"github.com/google/exposure-notifications-server/internal/mirror/model"
+	"github.com/jackc/pgx/v4"
+)
+
+type MirrorDB struct {
+	db *database.DB
+}
+
+func New(db *database.DB) *MirrorDB {
+	return &MirrorDB{
+		db: db,
+	}
+}
+
+func (db *MirrorDB) AddMirror(ctx context.Context, m *model.Mirror) error {
+	return db.db.InTx(ctx, pgx.Serializable, func(tx pgx.Tx) error {
+		row := tx.QueryRow(ctx, `
+			INSERT INTO
+				Mirror (index_file, export_root, cloud_storage_bucket, filename_root)
+			VALUES
+				($1, $2, $3, $4)
+			RETURNING id
+		`, m.IndexFile, m.ExportRoot, m.CloudStorageBucket, m.FilenameRoot)
+
+		if err := row.Scan(&m.ID); err != nil {
+			return fmt.Errorf("fetching mirror.ID: %w", err)
+		}
+		return nil
+	})
+}
+
+func (db *MirrorDB) DeleteMirror(ctx context.Context, m *model.Mirror) error {
+	return db.db.InTx(ctx, pgx.Serializable, func(tx pgx.Tx) error {
+		_, err := tx.Exec(ctx, `
+			DELETE FROM
+				MirrorFile
+			WHERE
+				mirror_id = $1
+			`, m.ID)
+		if err != nil {
+			return fmt.Errorf("failed to delete mirror files: %w", err)
+		}
+
+		_, err = tx.Exec(ctx, `
+			DELETE FROM
+				Mirror
+			WHERE
+				id = $1
+			`, m.ID)
+		if err != nil {
+			return fmt.Errorf("failed to delete mirror config: %w", err)
+		}
+
+		return nil
+	})
+}
+
+func (db *MirrorDB) Mirrors(ctx context.Context) ([]*model.Mirror, error) {
+	var mirrors []*model.Mirror
+
+	if err := db.db.InTx(ctx, pgx.ReadCommitted, func(tx pgx.Tx) error {
+		rows, err := tx.Query(ctx, `
+			SELECT
+				id, index_file, export_root, cloud_storage_bucket, filename_root
+			FROM
+				mirror
```

Comment:
`ORDER BY id` or something?

Reply:
added, but not necessary for proper function.